preview/Win7Msix/Win7MSIXInstaller/RegistryDevirtualizer.cpp

```diff
@@ -1,412 +1,412 @@
+        {
+            TraceLoggingWrite(g_MsixTraceLoggingProvider,
+                "Key not found -- nothing to delete",
+                TraceLoggingValue(realKey->GetPath().c_str(), "Realkey"),
+                TraceLoggingValue(subKeyName, "SubKey"));
+            return S_OK;
+        }
+        RETURN_IF_FAILED(hrCreateKey);
+
+        RETURN_IF_FAILED(RemoveDevirtualizeRegistryTree(&sourceSubKey, &destinationSubKey));
+
+        destinationSubKey.Close();
+        RETURN_IF_FAILED(RemoveSubKeyIfEmpty(realKey, subKeyName));
+        return S_OK;
+    }));
+
+    DWORD valuesCount = 0;
+    DWORD valueNameMaxLength = 0;
+    DWORD valueDataMaxLength = 0;
+    RETURN_IF_FAILED(virtualKey->GetValuesInfo(&valuesCount, &valueNameMaxLength, &valueDataMaxLength));
+
+    if (valuesCount > 0)
+    {
+        DWORD valueType = REG_NONE;
+        std::vector<wchar_t> valueNameBuffer(valueNameMaxLength + 1);
+        std::vector<BYTE> valueDataBuffer((valueDataMaxLength + 1));
+
+        // for each value of this key, devirtualize it and delete it from the real key.
+        for (DWORD i = 0; i < valuesCount; i++)
+        {
+            DWORD nameLength = static_cast<DWORD>(valueNameBuffer.size());
+            DWORD dataLength = static_cast<DWORD>(valueDataBuffer.size());
+            valueType = REG_NONE;
+
+            RETURN_IF_FAILED(virtualKey->EnumValue(i, &valueNameBuffer[0], &nameLength, &valueType,
+                reinterpret_cast<LPBYTE>(&valueDataBuffer[0]), &dataLength));
+
+            std::wstring valueNameString(valueNameBuffer.begin(), valueNameBuffer.end());
+            RETURN_IF_FAILED(realKey->DeleteValue(valueNameString.c_str()));
+        }
+    }
+
+    return S_OK;
+}
+
+HRESULT RegistryDevirtualizer::Create(std::wstring hiveFileName, MsixRequest* msixRequest, RegistryDevirtualizer ** instance)
+{
+    std::unique_ptr<RegistryDevirtualizer> localInstance(new RegistryDevirtualizer(hiveFileName, msixRequest));
+    if (localInstance == nullptr)
+    {
+        RETURN_IF_FAILED(E_OUTOFMEMORY);
+    }
+    
+    HANDLE userToken = nullptr;
+    if (!OpenProcessToken(GetCurrentProcess(), TOKEN_QUERY | TOKEN_ADJUST_PRIVILEGES, &userToken))
+    {
+        RETURN_IF_FAILED(HRESULT_FROM_WIN32(GetLastError()));
+    }
+
+    LUID seRestoreLuid;
+
+    if (!LookupPrivilegeValue(
+        NULL,            // lookup privilege on local system
+        SE_RESTORE_NAME,   // privilege to lookup 
+        &seRestoreLuid))        // receives LUID of privilege
+    {
+        RETURN_IF_FAILED(HRESULT_FROM_WIN32(GetLastError()));
+    }
+
+    LUID seBackupLuid;
+
+    if (!LookupPrivilegeValue(
+        NULL,            // lookup privilege on local system
+        SE_BACKUP_NAME,   // privilege to lookup 
+        &seBackupLuid))        // receives LUID of privilege
+    {
+        RETURN_IF_FAILED(HRESULT_FROM_WIN32(GetLastError()));
+    }
+
+    PTOKEN_PRIVILEGES pTokenPrivileges = NULL;
+    TOKEN_PRIVILEGES oldTokenPrivileges{};
```

Comment:
I noticed that I had to modify this such that, both pTokenPrivileges and oldTokenPrivileges are allocated in the same way initially for app installation to work.

Reply:
`oldTokenPrivileges` is never used, we should remove it.